Badly formatted response from Connect can cause the operator to get stack on a single reconciliation

In the `KafkaConnectApiImpl` class, when we get some _unexpected_ response from Kafka Connect REST API, we try to decode it to get a user-friendly error message. However, in some cases, it seems like Connect does not return a proper JSON. That causes the decoding to fail, but the exception is not caught and the reconciliation never completes:

```
2022-06-17 01:04:01 INFO  AbstractConnectOperator:223 - Reconciliation #2(connector-watch) KafkaConnect(kafka-connect/connect-cluster-test): KafkaConnector xxx in namespace kafka-connect was ADDED
2022-06-17 01:04:01 INFO  AbstractConnectOperator:445 - Reconciliation #2(connector-watch) KafkaConnect(kafka-connect/connect-cluster-test): creating/updating connector: xxx
...
2022-06-17 01:04:02 ERROR ContextImpl: - Unhandled exception
io.vertx.core.json.DecodeException: Failed to decode:No content to map due to end-of-input
 at [Source: (io.netty.buffer.ByteBufInputStream); line: 1, column: 0]
	at io.vertx.core.json.jackson.DatabindCodec.fromParser(DatabindCodec.java:129) ~[io.vertx.vertx-core-4.2.4.jar:4.2.4]
	at io.vertx.core.json.jackson.DatabindCodec.fromBuffer(DatabindCodec.java:99) ~[io.vertx.vertx-core-4.2.4.jar:4.2.4]
	at io.vertx.core.json.JsonObject.fromBuffer(JsonObject.java:948) ~[io.vertx.vertx-core-4.2.4.jar:4.2.4]
	at io.vertx.core.json.JsonObject.<init>(JsonObject.java:85) ~[io.vertx.vertx-core-4.2.4.jar:4.2.4]
	at io.vertx.core.buffer.impl.BufferImpl.toJsonObject(BufferImpl.java:110) ~[io.vertx.vertx-core-4.2.4.jar:4.2.4]
	at io.strimzi.operator.cluster.operator.assembly.KafkaConnectApiImpl.lambda$createOrUpdatePutRequest$1(KafkaConnectApiImpl.java:95) ~[io.strimzi.cluster-operator-0.28.0.jar:0.28.0]
	at io.vertx.core.impl.future.FutureImpl$1.onSuccess(FutureImpl.java:91) ~[io.vertx.vertx-core-4.2.4.jar:4.2.4]
	at io.vertx.core.impl.future.FutureBase.lambda$emitSuccess$0(FutureBase.java:54) ~[io.vertx.vertx-core-4.2.4.jar:4.2.4]
	at io.vertx.core.impl.EventLoopContext.execute(EventLoopContext.java:81) ~[io.vertx.vertx-core-4.2.4.jar:4.2.4]
	at io.vertx.core.impl.ContextImpl.execute(ContextImpl.java:260) ~[io.vertx.vertx-core-4.2.4.jar:4.2.4]
	at io.vertx.core.impl.EventLoopContext.execute(EventLoopContext.java:22) ~[io.vertx.vertx-core-4.2.4.jar:4.2.4]
	at io.vertx.core.impl.future.FutureBase.emitSuccess(FutureBase.java:51) ~[io.vertx.vertx-core-4.2.4.jar:4.2.4]
	at io.vertx.core.impl.future.FutureImpl.tryComplete(FutureImpl.java:211) ~[io.vertx.vertx-core-4.2.4.jar:4.2.4]
	at io.vertx.core.impl.future.PromiseImpl.tryComplete(PromiseImpl.java:23) ~[io.vertx.vertx-core-4.2.4.jar:4.2.4]
	at io.vertx.core.http.impl.HttpEventHandler.handleEnd(HttpEventHandler.java:79) ~[io.vertx.vertx-core-4.2.4.jar:4.2.4]
	at io.vertx.core.http.impl.HttpClientResponseImpl.handleEnd(HttpClientResponseImpl.java:250) ~[io.vertx.vertx-core-4.2.4.jar:4.2.4]
	at io.vertx.core.http.impl.Http1xClientConnection$StreamImpl.lambda$new$0(Http1xClientConnection.java:383) ~[io.vertx.vertx-core-4.2.4.jar:4.2.4]
	at io.vertx.core.streams.impl.InboundBuffer.handleEvent(InboundBuffer.java:240) ~[io.vertx.vertx-core-4.2.4.jar:4.2.4]
	at io.vertx.core.streams.impl.InboundBuffer.write(InboundBuffer.java:130) ~[io.vertx.vertx-core-4.2.4.jar:4.2.4]
	at io.vertx.core.http.impl.Http1xClientConnection$StreamImpl.handleEnd(Http1xClientConnection.java:610) ~[io.vertx.vertx-core-4.2.4.jar:4.2.4]
	at io.vertx.core.impl.EventLoopContext.execute(EventLoopContext.java:71) [io.vertx.vertx-core-4.2.4.jar:4.2.4]
	at io.vertx.core.impl.ContextImpl.execute(ContextImpl.java:267) [io.vertx.vertx-core-4.2.4.jar:4.2.4]
	at io.vertx.core.http.impl.Http1xClientConnection.handleResponseEnd(Http1xClientConnection.java:844) [io.vertx.vertx-core-4.2.4.jar:4.2.4]
	at io.vertx.core.http.impl.Http1xClientConnection.handleHttpMessage(Http1xClientConnection.java:716) [io.vertx.vertx-core-4.2.4.jar:4.2.4]
	at io.vertx.core.http.impl.Http1xClientConnection.handleMessage(Http1xClientConnection.java:680) [io.vertx.vertx-core-4.2.4.jar:4.2.4]
	at io.vertx.core.net.impl.ConnectionBase.read(ConnectionBase.java:156) [io.vertx.vertx-core-4.2.4.jar:4.2.4]
	at io.vertx.core.net.impl.VertxHandler.channelRead(VertxHandler.java:153) [io.vertx.vertx-core-4.2.4.jar:4.2.4]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379) [io.netty.netty-transport-4.1.71.Final.jar:4.1.71.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365) [io.netty.netty-transport-4.1.71.Final.jar:4.1.71.Final]
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357) [io.netty.netty-transport-4.1.71.Final.jar:4.1.71.Final]
	at io.netty.channel.CombinedChannelDuplexHandler$DelegatingChannelHandlerContext.fireChannelRead(CombinedChannelDuplexHandler.java:436) [io.netty.netty-transport-4.1.71.Final.jar:4.1.71.Final]
	at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:324) [io.netty.netty-codec-4.1.71.Final.jar:4.1.71.Final]
	at io.netty.handler.codec.ByteToMessageDecoder.channelInputClosed(ByteToMessageDecoder.java:383) [io.netty.netty-codec-4.1.71.Final.jar:4.1.71.Final]
	at io.netty.handler.codec.ByteToMessageDecoder.channelInactive(ByteToMessageDecoder.java:354) [io.netty.netty-codec-4.1.71.Final.jar:4.1.71.Final]
	at io.netty.handler.codec.http.HttpClientCodec$Decoder.channelInactive(HttpClientCodec.java:326) [io.netty.netty-codec-http-4.1.71.Final.jar:4.1.71.Final]
	at io.netty.channel.CombinedChannelDuplexHandler.channelInactive(CombinedChannelDuplexHandler.java:221) [io.netty.netty-transport-4.1.71.Final.jar:4.1.71.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:262) [io.netty.netty-transport-4.1.71.Final.jar:4.1.71.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:248) [io.netty.netty-transport-4.1.71.Final.jar:4.1.71.Final]
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelInactive(AbstractChannelHandlerContext.java:241) [io.netty.netty-transport-4.1.71.Final.jar:4.1.71.Final]
	at io.netty.handler.logging.LoggingHandler.channelInactive(LoggingHandler.java:206) [io.netty.netty-handler-4.1.71.Final.jar:4.1.71.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:262) [io.netty.netty-transport-4.1.71.Final.jar:4.1.71.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:248) [io.netty.netty-transport-4.1.71.Final.jar:4.1.71.Final]
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelInactive(AbstractChannelHandlerContext.java:241) [io.netty.netty-transport-4.1.71.Final.jar:4.1.71.Final]
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelInactive(DefaultChannelPipeline.java:1405) [io.netty.netty-transport-4.1.71.Final.jar:4.1.71.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:262) [io.netty.netty-transport-4.1.71.Final.jar:4.1.71.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:248) [io.netty.netty-transport-4.1.71.Final.jar:4.1.71.Final]
	at io.netty.channel.DefaultChannelPipeline.fireChannelInactive(DefaultChannelPipeline.java:901) [io.netty.netty-transport-4.1.71.Final.jar:4.1.71.Final]
	at io.netty.channel.AbstractChannel$AbstractUnsafe$8.run(AbstractChannel.java:831) [io.netty.netty-transport-4.1.71.Final.jar:4.1.71.Final]
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:164) [io.netty.netty-common-4.1.71.Final.jar:4.1.71.Final]
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:469) [io.netty.netty-common-4.1.71.Final.jar:4.1.71.Final]
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:500) [io.netty.netty-transport-4.1.71.Final.jar:4.1.71.Final]
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:986) [io.netty.netty-common-4.1.71.Final.jar:4.1.71.Final]
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74) [io.netty.netty-common-4.1.71.Final.jar:4.1.71.Final]
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30) [io.netty.netty-common-4.1.71.Final.jar:4.1.71.Final]
	at java.lang.Thread.run(Thread.java:829) [?:?]
Caused by: com.fasterxml.jackson.databind.exc.MismatchedInputException: No content to map due to end-of-input
 at [Source: (io.netty.buffer.ByteBufInputStream); line: 1, column: 0]
	at com.fasterxml.jackson.databind.exc.MismatchedInputException.from(MismatchedInputException.java:59) ~[com.fasterxml.jackson.core.jackson-databind-2.11.3.jar:2.11.3]
	at com.fasterxml.jackson.databind.ObjectMapper._initForReading(ObjectMapper.java:4668) ~[com.fasterxml.jackson.core.jackson-databind-2.11.3.jar:2.11.3]
	at com.fasterxml.jackson.databind.ObjectMapper._readValue(ObjectMapper.java:4484) ~[com.fasterxml.jackson.core.jackson-databind-2.11.3.jar:2.11.3]
	at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:2730) ~[com.fasterxml.jackson.core.jackson-databind-2.11.3.jar:2.11.3]
	at io.vertx.core.json.jackson.DatabindCodec.fromParser(DatabindCodec.java:126) ~[io.vertx.vertx-core-4.2.4.jar:4.2.4]
	... 54 more
...
2022-06-17 01:05:01 INFO  AbstractOperator:373 - Reconciliation #2(connector-watch) KafkaConnect(kafka-connect/connect-cluster-test): Reconciliation is in progress
...
2022-06-17 01:06:01 INFO  AbstractOperator:373 - Reconciliation #2(connector-watch) KafkaConnect(kafka-connect/connect-cluster-test): Reconciliation is in progress
...
2022-06-17 01:07:01 INFO  AbstractOperator:373 - Reconciliation #2(connector-watch) KafkaConnect(kafka-connect/connect-cluster-test): Reconciliation is in progress
...
2022-06-17 01:08:01 INFO  AbstractOperator:373 - Reconciliation #2(connector-watch) KafkaConnect(kafka-connect/connect-cluster-test): Reconciliation is in progress
...
```

This seems to be coming from here: https://github.com/strimzi/strimzi-kafka-operator/blob/ca13920bed1c3eb1e8e08951f06936968413f576/cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaConnectApiImpl.java#L95

But the same code is on few more places as well and shoud be fixed everywhere.

At this point, it is not clear what does Connect return and why, but the code is clearly vulnerable to this and can be improved. Based on https://cloud-native.slack.com/archives/CMH3Q3SNP/p1655468396285749

--- a/cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaConnectApiImpl.java
+++ b/cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaConnectApiImpl.java
@@ -36,6 +36,7 @@ import io.vertx.core.Vertx;
 import io.vertx.core.buffer.Buffer;
 import io.vertx.core.http.HttpClientOptions;
 import io.vertx.core.http.HttpMethod;
+import io.vertx.core.json.DecodeException;
 import io.vertx.core.json.JsonArray;
 import io.vertx.core.json.JsonObject;
 
@@ -92,8 +93,7 @@ class KafkaConnectApiImpl implements KafkaConnectApi {
                                 // TODO Handle 409 (Conflict) indicating a rebalance in progress
                                 LOGGER.debugCr(reconciliation, "Got {} response to PUT request to {}", response.result().statusCode(), path);
                                 response.result().bodyHandler(buffer -> {
-                                    JsonObject x = buffer.toJsonObject();
-                                    result.fail(new ConnectRestException(response.result(), x.getString("message")));
+                                    result.fail(new ConnectRestException(response.result(), tryToExtractErrorMessage(reconciliation, buffer)));
                                 });
                             }
                         } else {
@@ -139,8 +139,7 @@ class KafkaConnectApiImpl implements KafkaConnectApi {
                                 // TODO Handle 409 (Conflict) indicating a rebalance in progress
                                 LOGGER.debugCr(reconciliation, "Got {} response to GET request to {}", response.result().statusCode(), path);
                                 response.result().bodyHandler(buffer -> {
-                                    JsonObject x = buffer.toJsonObject();
-                                    result.fail(new ConnectRestException(response.result(), x.getString("message")));
+                                    result.fail(new ConnectRestException(response.result(), tryToExtractErrorMessage(reconciliation, buffer)));
                                 });
                             }
                         } else {
@@ -193,9 +192,9 @@ class KafkaConnectApiImpl implements KafkaConnectApi {
                                     });
                             } else {
                                 // TODO Handle 409 (Conflict) indicating a rebalance in progress
+                                LOGGER.debugCr(reconciliation, "Got {} response to DELETE request to {}", response.result().statusCode(), path);
                                 response.result().bodyHandler(buffer -> {
-                                    JsonObject x = buffer.toJsonObject();
-                                    result.fail(new ConnectRestException(response.result(), x.getString("message")));
+                                    result.fail(new ConnectRestException(response.result(), tryToExtractErrorMessage(reconciliation, buffer)));
                                 });
                             }
                         } else {
@@ -625,5 +624,15 @@ class KafkaConnectApiImpl implements KafkaConnectApi {
                     result.tryFail(request.cause());
                 }
             }));
+    }
+
+    /* test */ static String tryToExtractErrorMessage(Reconciliation reconciliation, Buffer buffer)    {
+        try {
+            return buffer.toJsonObject().getString("message");
+        } catch (DecodeException e) {
+            LOGGER.warnCr(reconciliation, "Failed to decode the error message from the response: " + buffer, e);
+        }
+
+        return "Unknown error message";
     }
 }

--- a/cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/KafkaConnectApiImplTest.java
+++ b/cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/KafkaConnectApiImplTest.java
@@ -1,0 +1,70 @@
+/*
+ * Copyright Strimzi authors.
+ * License: Apache License 2.0 (see the file LICENSE or http://apache.org/licenses/LICENSE-2.0.html).
+ */
+package io.strimzi.operator.cluster.operator.assembly;
+
+import io.strimzi.operator.common.Reconciliation;
+import io.vertx.core.Vertx;
+import io.vertx.core.buffer.Buffer;
+import io.vertx.core.http.HttpServer;
+import io.vertx.core.json.JsonObject;
+import io.vertx.junit5.Checkpoint;
+import io.vertx.junit5.VertxExtension;
+import io.vertx.junit5.VertxTestContext;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+
+import java.util.concurrent.ExecutionException;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.containsString;
+import static org.hamcrest.Matchers.is;
+
+@ExtendWith(VertxExtension.class)
+public class KafkaConnectApiImplTest {
+    @Test
+    public void testJsonDecoding()  {
+        assertThat(KafkaConnectApiImpl.tryToExtractErrorMessage(Reconciliation.DUMMY_RECONCILIATION, Buffer.buffer("{\"message\": \"This is the error\"}")), is("This is the error"));
+        assertThat(KafkaConnectApiImpl.tryToExtractErrorMessage(Reconciliation.DUMMY_RECONCILIATION, Buffer.buffer("{\"message\": \"This is the error\"")), is("Unknown error message"));
+        assertThat(KafkaConnectApiImpl.tryToExtractErrorMessage(Reconciliation.DUMMY_RECONCILIATION, Buffer.buffer("Not a JSON")), is("Unknown error message"));
+    }
+
+    @Test
+    public void testFeatureCompletionWithBadlyFormattedError(Vertx vertx, VertxTestContext context) throws ExecutionException, InterruptedException {
+        HttpServer server = mockApi(vertx, "Some error message");
+
+        KafkaConnectApi api = new KafkaConnectApiImpl(vertx);
+
+        Checkpoint async = context.checkpoint();
+        api.createOrUpdatePutRequest(Reconciliation.DUMMY_RECONCILIATION, "127.0.0.1", server.actualPort(), "my-connector", new JsonObject())
+                        .onComplete(context.failing(res -> context.verify(() -> {
+                            assertThat(res.getMessage(), containsString("Unknown error message"));
+
+                            server.close();
+                            async.flag();
+                        })));
+    }
+
+    @Test
+    public void testFeatureCompletionWithWellFormattedError(Vertx vertx, VertxTestContext context) throws ExecutionException, InterruptedException {
+        HttpServer server = mockApi(vertx, "{\"message\": \"This is the error\"}");
+
+        KafkaConnectApi api = new KafkaConnectApiImpl(vertx);
+
+        Checkpoint async = context.checkpoint();
+        api.createOrUpdatePutRequest(Reconciliation.DUMMY_RECONCILIATION, "127.0.0.1", server.actualPort(), "my-connector", new JsonObject())
+                .onComplete(context.failing(res -> context.verify(() -> {
+                    assertThat(res.getMessage(), containsString("This is the error"));
+
+                    server.close();
+                    async.flag();
+                })));
+    }
+
+    HttpServer mockApi(Vertx vertx, String error) throws InterruptedException, ExecutionException {
+        HttpServer httpServer = vertx.createHttpServer().requestHandler(request -> request.response().setStatusCode(500).end(error));
+
+        return httpServer.listen(0).toCompletionStage().toCompletableFuture().get();
+    }
+}

